### PR TITLE
[autoscaler] Create Docker Command Runner

### LINKED ISSUE #4183: [autoscaler] Ray submit with docker doesn't rsync the script to the container
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Debian GNU/Linux 9.8 (stretch) on GCP
- **Ray installed from (source or binary)**: source (5bfcfa8)
- **Ray version**: master - 5bfcfa8 (0.7.0.dev0)
- **Python version**: 3.7
- **Exact command to reproduce**:
```
ray up -y minimal.yaml  # with a docker image of the master build of Ray
ray submit --docker minimal.yaml hello_ray.py
```

### Describe the problem
If I run `submit` without `--docker` then the worker can execute the specified script from `~`, but when `--docker` is set, it throws:
```
python: can't open file '/root/hello_ray.py': [Errno 2] No such file or directory
```

When the `docker run` is called in the initialization phase, there doesn't seem to be any attached volumes, so I guess a shared location between the host and the container should resolve this, but there could be better ways as well. 

### Source code / logs
I can include full logs if necessary.

### LINKED ISSUE #4403: [tune/autoscaler] _LogSyncer cannot rsync with Docker
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 18.04
- **Ray installed from (source or binary)**: binary
- **Ray version**: 0.6.4
- **Python version**: 3.6.7

### Describe the problem
In `_LogSyncer`, `sync_to_worker_if_possible` and `sync_now` use `rsync` to transfer logs between the local node and the worker. This breaks when using Docker, since:
  - If the local node is in Docker, it will typically have the `root` username, and so this is what `get_ssh_user` will return. But we cannot typically login to the worker node as `root`.
  - The `local_dir` on the worker is inside the Docker container, and may not even be visible outside. If it is bound, then it will typically be at a different path.

An unrelated issue: if `self.sync_func` is non-None, it will get executed *before* the `worker_to_local_sync_cmd`, which I think is wrong.

I'd be happy to make a stab at a PR, but I'd appreciate some suggestions on the right way of fixing this, as it's been a while since I've looked at Ray internals. This also feels like a problem that is likely to reoccur with slight variation, e.g. this bug is similar to #4183

Perhaps we can make `autoscaler` provide an abstract `sync` interface that `tune` and other consumers can use. This could make to `rsync` in the standard case, and something more complex in the Docker case (e.g. Docker cp followed by rsync)? `ray.autoscaller.commands.rsync` is already something along these lines -- would this be an appropriate place to modify?

A more hacky solution would be to make `get_ssh_user` return the right value and make the Docker volume-binding line up so that we can just ignore the difference between Docker and non-Docker instances.

### Source code / logs
A MWE for this is hard to provide, but if the above description is insufficient I can try to come up with one.

### PR DESCRIPTION
**Pull request by ijrsvt**
## Why are these changes needed?

This PR is the first in a few PRs to refactor how the autoscaler interfaces with docker. This PR mainly focuses on how the CommandRunner is instantiated. The DockerCommandRunner mostly passes commands down to an enclosed SSHCommandRunner. The main functionality change
 is that rsync actually sends the file into docker.

## Related issue number

Closes #4183
Closes #4403
Replaces #8527

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/autoscaler/updater.py [hunk 1]
```diff
@@ -312,14 +353,15 @@ def __init__(self,
                  ray_start_commands,
                  runtime_hash,
                  process_runner=subprocess,
-                 use_internal_ip=False):
+                 use_internal_ip=False,
+                 docker_config=None):
```
Comment (ijrsvt) [inline python/ray/autoscaler/updater.py:357]: Should we just pass the entire config in?
Currently `provider_config`, `auth_config`, `cluster_name`, and `file_mounts` (and now `docker_config`) are all parts

File: python/ray/autoscaler/updater.py [hunk 2]
```diff
@@ -297,6 +297,47 @@ def remote_shell_command_str(self):
...
+
+    def remote_shell_command_str(self):
+        inner_str = self.ssh_command_runner.remote_shell_command_str().replace(
+            "ssh", "ssh -tt", 1).strip("\n")
+        return inner_str + " docker exec -it {} /bin/bash\n".format(
+            self.docker_name)
+
+    def clean_squiggly(self, string):
+        return string.replace(
+            "~", "`docker exec ray_docker env | grep HOME | cut -d'=' -f2`")
```
Comment (richardliaw) [inline python/ray/autoscaler/updater.py:338]: I think this should only be in the front of the path. For example, if my path is "user/~file.txt", we shouldn't replace it.

Also, call it `docker_expand_user`?

File: python/ray/autoscaler/updater.py [hunk 3]
```diff
@@ -297,6 +297,47 @@ def remote_shell_command_str(self):
             self.ssh_private_key, self.ssh_user, self.ssh_ip)
 
 
+class DockerCommandRunner(SSHCommandRunner):
+    def __init__(self, doker_config, log_prefix, node_id, provider,
```
Comment (richardliaw) [inline python/ray/autoscaler/updater.py:301]: docker

File: python/ray/autoscaler/updater.py (hunk 1 above)
Comment (richardliaw) [reply python/ray/autoscaler/updater.py:357]: No, let's pass in the docker_config like what you're doing now

File: python/ray/autoscaler/node_provider.py [hunk 4]
```diff
@@ -226,7 +232,14 @@ def get_command_runner(self, log_prefix, node_id, auth_config,
...
-        return SSHCommandRunner(log_prefix, node_id, self, auth_config,
-                                cluster_name, process_runner, use_internal_ip)
+        if docker_config and docker_config["container_name"] != "":
+            return DockerCommandRunner(docker_config, log_prefix, node_id,
+                                       self, auth_config, cluster_name,
+                                       process_runner, use_internal_ip)
+        else:
+            return SSHCommandRunner(log_prefix, node_id, self, auth_config,
+                                    cluster_name, process_runner,
+                                    use_internal_ip)
```
Comment (richardliaw) [inline python/ray/autoscaler/node_provider.py:245]: Can you build the args with a dictionary? 

```
common_args = {...}
if should_use_docker(docker_config):
    return DockerCommandRunner(docker_config=docker_config, **common_args)
...
```

File: python/ray/autoscaler/updater.py (hunk 2 above)
Comment (ijrsvt) [reply python/ray/autoscaler/updater.py:338]: Oh--I did not realize that is a valid string. I will change it to `docker_expand_user`

File: python/ray/autoscaler/node_provider.py (hunk 4 above)
Comment (ijrsvt) [reply python/ray/autoscaler/node_provider.py:245]: Oh--sure. I didn't realize I could do that!

File: python/ray/autoscaler/updater.py (hunk 1 above)
Comment (ijrsvt) [reply python/ray/autoscaler/updater.py:358]: Okay--is there a specific reason for this?

**Review by ericl — COMMENTED** [review]
Looks nice!

File: python/ray/autoscaler/updater.py (hunk 1 above)
Comment (richardliaw) [reply python/ray/autoscaler/updater.py:358]: Well, it helps isolate what is needed and what is not needed. For example, initialization_commands is also part of the config, but there are times when we use the NodeUpdater where we want to run it and when we don't want to.

**Comment by richardliaw** [comment]
Is this ready to merge?

**Comment by ijrsvt** [comment]
@richardliaw  Yes it is!

**Review by richardliaw — APPROVED** [review]

**Pull request merged**